### PR TITLE
Make PixelData.get pixels more efficient

### PR DESCRIPTION
**Pull request by hsautessella**
This PR improves the performance of `PixelData.get_pixels`. Previously we were reading in whole pages and discarding pixels we didn't need. This new implementation reads only the required pixels, which means we're not doing more file reading than we need to.

Addresses #477

### CODE REVIEW

File: _test/test_sqw/test_PixelData.m [hunk 1]
```diff
@@ -1497,6 +1510,77 @@ function test_base_page_size_is_DEFAULT_PAGE_SIZE_by_default(~)
...
+        % relevant test explanation
+        pix = PixelData(obj.test_sqw_file_path, obj.SMALL_PG_SIZE);
+        pix.advance();
+        pix.signal = 11;
+        pix.advance();
+        pix.signal = 12;
+        pix.advance();
+
+        pg_size = pix.base_page_size;
+        pix_range = pix.num_pixels:-1:1;
```
Comment (nickbattam-tessella) [inline _test/test_sqw/test_PixelData.m:1552]: add a comment to this emphasising that your pix_range is reversed

File: horace_core/sqw/PixelData/@PixelData/get_pixels.m [hunk 2]
```diff
@@ -30,19 +30,34 @@
...
+        [pix_out, dirty_pg_mask] = ...
+            assign_dirty_pixels(obj, pix_out, abs_pix_indices);
+
+        % Now assign the clean pixels
+        [unique_sorted, ~, idx_map] = unique(abs_pix_indices(~dirty_pg_mask));
+        raw_pix = obj.f_accessor_.get_pix_at_indices(unique_sorted);
+        pix_out.data(:, ~dirty_pg_mask) = raw_pix(:, idx_map);
+
+    else
+        % All pixels in file
```
Comment (nickbattam-tessella) [inline horace_core/sqw/PixelData/@PixelData/get_pixels.m:49]: All pixels are in file for the temp file case too -- this is all pixels are in the original sqw file

File: horace_core/sqw/PixelData/@PixelData/get_pixels.m [hunk 3]
```diff
@@ -30,19 +30,34 @@
...
+        if issorted(abs_pix_indices, 'strictascend')
```
Comment (nickbattam-tessella) [inline horace_core/sqw/PixelData/@PixelData/get_pixels.m:50]: Is it worth adding a comment noting that the sort/unsort code is significantly slower, even if the data is sorted? That explains why we are doing this check.

File: horace_core/sqw/PixelData/@PixelData/get_pixels.m [hunk 4]
```diff
@@ -30,19 +30,34 @@
...
+            pix_out = PixelData(obj.f_accessor_.get_pix_at_indices(abs_pix_indices));
+        else
+            % get_pix_at_indices requires monotonically increasing indices
```
Comment (nickbattam-tessella) [inline horace_core/sqw/PixelData/@PixelData/get_pixels.m:53]: this code block (except the PixelData object creation at the end) is repeated twice - is there a way to extract it as a function?

File: horace_core/sqw/PixelData/@PixelData/private/get_pg_idx_from_absolute_.m [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+%
+% Output:
+% -------
+%  page_idx    The page indices corresponding to the given absolute indices
+%              that exist within the given page.
+%  global_idx  The absolute indices that exist within the given page. This
+%              will be a subset of `abs_indices`.
+%
+pg_start_idx = (page_number - 1)*obj.base_page_size + 1;
+pg_end_idx = pg_start_idx + obj.base_page_size - 1;
```
Comment (nickbattam-tessella) [inline horace_core/sqw/PixelData/@PixelData/private/get_pg_idx_from_absolute_.m:29]: if you change the `<=` to `<`  you can drop the `-1`  which may reduce the complexity for anyone reading the code

File: horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m [hunk 6]
```diff
@@ -51,6 +51,48 @@
...
+
+        raw_pix = zeros(ncols, numel(indices));
+
+        num_pix_read = 0;
+        for block_num = 1:numel(read_sizes)
+            fseek(file_id, seek_sizes(block_num)*PIXEL_SIZE, 'cof');
+
+            out_pix_start = num_pix_read + 1;
+            out_pix_end = out_pix_start + read_sizes(block_num) - 1;
+            read_pix = fread(file_id, [ncols, read_sizes(block_num)], obj.FILE_DATA_FORMAT_);
```
Comment (nickbattam-tessella) [inline horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m:86]: there's no error handling with this `fread` or `fseek` - can you reuse `do_fread` and `do_fseek` defined (currently) as part of `sqw_binfile_common`?
These are generic file reading helpers so could reasonably move into the utilities folder

File: horace_core/sqw/PixelData/@PixelData/get_pixels.m [hunk 7]
```diff
@@ -81,3 +96,51 @@
 function is = is_positive_int_vector_or_logical_vector(vec)
     is = isvector(vec) && (islogical(vec) || (all(vec > 0 & all(floor(vec) == vec))));
 end
+
+
+function [pix, dirty_pg_mask] = assign_dirty_pixels(obj, pix, abs_pix_indices)
```
Comment (nickbattam-tessella) [inline horace_core/sqw/PixelData/@PixelData/get_pixels.m:101]: `assign_` or `read_`?

File: horace_core/sqw/file_io/@sqw_binfile_common/get_pix_at_indices.m [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+if ~obj.is_activated('read')
+    obj = obj.activate('read');
+end
+
+NUM_BYTES_IN_FLOAT = 4;
+PIXEL_SIZE = NUM_BYTES_IN_FLOAT*PixelData.DEFAULT_NUM_PIX_FIELDS;  % bytes
+
+[read_sizes, seek_sizes] = get_read_and_seek_sizes(indices(:)');
+
+% Pre-allocate output array
```
Comment (abuts) [inline horace_core/sqw/file_io/@sqw_binfile_common/get_pix_at_indices.m:22]:
```suggestion
% Pre-allocate output array
% Pre-allocate output array
pix = zeros(PixelData.DEFAULT_NUM_PIX_FIELDS, numel(indices));

% Position file reader at start of pixel array
do_fseek(obj.file_id_, obj.pix_pos_, 'bof');

num_pix_read = 0;
for block_num = 1:numel(read_sizes)
    do_fseek(obj.file_id_, seek_sizes(block_num)*PIXEL_SIZE, 'cof');

    out_pix_start = num_pix_read + 1;
    out_pix_end = out_pix_start + read_sizes(block_num) - 1;
    pix(:, out_pix_start:out_pix_end) = ...
        do_fread(obj.file_id_, read_sizes(block_num));

    num_pix_read = num_pix_read + read_sizes(block_num);
end
```
I believe 
```
block_tmp = cell(1,num_blocks)
...
block_tmp{block_num} = read(block(block_num));
...
pix = [block_tmp{:}]; % may need proper sizing 
```
will work much faster in Matlab

**Review by nickbattam-tessella — APPROVED** [review]

**Pull request merged**